service_configuration_lib/spark_config.py

```diff
@@ -428,30 +431,81 @@ def compute_executor_instances_k8s(user_spark_opts: Dict[str, str]) -> int:
             f'Given executor memory is {memory_mb / 1024}g, '
             f'capped to {executor_memory} to better fit on available aws nodes.',
         )
-    elif memory_mb > EXECUTOR_MEMORY_WARN_GB * 1024:
-        log.warning('We recommend using setting memory as 28g and executor cores as 4')
+    elif memory_mb > RECOMMENDED_EXECUTOR_MEMORY * 1024:
+        log.warning(
+            f'We recommend using setting memory as {RECOMMENDED_EXECUTOR_MEMORY}g '
+            f'and executor cores as 4',
+        )
 
-    if executor_cores > EXECUTOR_CORES_CAP:
+    if executor_cores > MAX_EXECUTOR_CORES_ALLOWED:
         log.warning(
             f'Given executor cores is {executor_cores}, '
-            f'capped to {EXECUTOR_CORES_CAP} to better fit on available aws nodes.',
+            f'capped to {MAX_EXECUTOR_CORES_ALLOWED} to better fit on available aws nodes.',
         )
-        executor_cores = EXECUTOR_CORES_CAP
+        executor_cores = MAX_EXECUTOR_CORES_ALLOWED
+
+    return executor_cores, executor_memory
+
+
+def _recalculate_executor_resources(
+    user_spark_opts: Dict[str, str],
+    force_spark_resource_configs: bool,
+) -> Dict[str, str]:
+    executor_cores = int(user_spark_opts.get('spark.executor.cores', str(DEFAULT_EXECUTOR_CORES)))
+    executor_memory = user_spark_opts.get('spark.executor.memory', f'{DEFAULT_EXECUTOR_MEMORY}g')
+    executor_instances = int(user_spark_opts.get('spark.executor.instances', str(DEFAULT_EXECUTOR_INSTANCES)))
+
+    memory_mb = parse_memory_string(executor_memory)
+
+    if force_spark_resource_configs:
+        log.warning(
+            'force_spark_resource_configs is set to true: '
+            'this can result in non-optimal bin-packing of executors on aws nodes or '
+            'can lead to wastage the resources. '
+            'Please make sure there is a reason that this flag needs to be used.',
+        )
+        executor_cores, executor_memory = _cap_executor_resources(executor_cores, executor_memory, memory_mb)
+    elif (
+        memory_mb > MEDIUM_EXECUTOR_MEMORY * 1024 or
+        executor_cores > MEDIUM_EXECUTOR_CORES
+    ):
+        capped_executor_memory = f'{MEDIUM_EXECUTOR_MEMORY}g'
+        capped_executor_cores = MEDIUM_EXECUTOR_CORES
+        capped_executor_instances = max(
+            executor_instances * memory_mb // (1024 * MEDIUM_EXECUTOR_MEMORY),
+            executor_instances * executor_cores // MEDIUM_EXECUTOR_CORES,
+        )
+        log.warning(
+            f'Given executor {executor_cores}C {int(memory_mb / 1024)}g x{executor_instances} instances '
+            f'=> adjusted to {capped_executor_cores}C {capped_executor_memory} x{capped_executor_instances} instances '
+            f'to better fit on available aws nodes.',
+        )
+        (executor_cores, executor_memory, executor_instances) = (
+            capped_executor_cores, capped_executor_memory, capped_executor_instances,
+        )
+    elif memory_mb > RECOMMENDED_EXECUTOR_MEMORY * 1024:
```

Comment:
We also need an `else` condition for jobs with memory < RECOMMENDED_EXECUTOR_MEMORY, to adjust to (28g, 4) pair while reducing the instances as long as instances is >=1.
If adjusted_instances == 0, just adjust to 7:1 ratio (example: 7g, 1core, 1 instance, 21g, 3core, 1 instance)